command_line/expeditor.py

```diff
@@ -0,0 +1,213 @@
+from __future__ import absolute_import, division, print_function
+
+from itertools import groupby
+from tabulate import tabulate
+import copy
+from libtbx.phil import parse
+from dials.util import show_mail_on_error
+from dials.util.options import OptionParser, flatten_reflections, flatten_experiments
+from dials.array_family import flex
+from dxtbx.model.experiment_list import ExperimentList
+
+help_message = """
+
+Utility to wrangle experiments from indexing in preparation for refinement -
+will look to see that the sample is centred, that there are observations for
+the full range of every scan and rewrite the experiment lists if not.
+
+  dials.expeditor indexed.expt indexed.refl
+
+"""
+
+
+phil_scope = parse(
+    """
+output {
+  experiments = expedited.expt
+    .type = path
+    .help = "Filename for fixed experiment list"
+
+  reflections = expedited.refl
+    .type = path
+    .help = "Filename for fixed reflection list"
+}
+"""
+)
+
+
+def model_connectivity(experiments):
+    def model_connectivity_impl(experiments, model):
+        text = [""]
+        text.append("%s:" % model.capitalize())
+        models = getattr(experiments, "%ss" % model)()
+        rows = [[""] + [str(j) for j in range(len(models))]]
+        for j, e in enumerate(experiments):
+            row = ["Experiment %d" % j]
+            for m in models:
+                if getattr(e, model) is m:
+                    row.append("x")
+                else:
+                    row.append(".")
+            rows.append(row)
+        text.append(tabulate(rows, tablefmt="plain"))
+        return text
+
+    if len(experiments) == 1:
+        return ""
+
+    text = []
+    text.append("Experiment / Models")
+    text.extend(model_connectivity_impl(experiments, "detector"))
+    text.extend(model_connectivity_impl(experiments, "crystal"))
+    text.extend(model_connectivity_impl(experiments, "beam"))
+
+    return "\n".join(text)
+
+
+def select_scans_from_reflections(reflections, scan):
+    """Determine a list of valid scans where reflection shoeboxes are seen,
+    from within the bounds of the input scan."""
+
+    bbox = reflections["bbox"].parts()
+
+    z0, z1 = bbox[4], bbox[5]
+    i0, i1 = scan.get_array_range()
+    o0, o1 = scan.get_oscillation()
+
+    coverage = flex.int(i1 - i0, 0)
+
+    for _z0, _z1 in zip(z0, z1):
+        for j in range(_z0, _z1):
+            coverage[j] += 1
+
+    filled = (coverage > 0).iselection()
+
+    scans = []
+
+    for k, g in groupby(enumerate(filled), lambda n: n[0] - n[1]):
+        l = list(g)
+        s = copy.deepcopy(scan)
+        s.set_image_range((l[0][1] + 1, l[-1][1] + 1))
```

Comment:
The use of the variable name `l` combined with five uses of number `1` in this line makes it very hard to read. I would generally discourage the use of single latter variable names where possible.
In fact, PEP8 actively discourages the use of the variable name `l`: https://www.python.org/dev/peps/pep-0008/#names-to-avoid
> Never use the characters 'l' (lowercase letter el), 'O' (uppercase letter oh), or 'I' (uppercase letter eye) as single character variable names.